File: desktop/helper/lib/creative-constraints.js
'use strict';

const { canonicalText } = require('./story-plan');

const CREATIVE_CONSTRAINTS_SCHEMA = 'autoeditor-creative-constraints/v1';
const GRAPHIC_KINDS = Object.freeze(['bars', 'callout', 'keyword', 'stat']);
const GRAPHIC_KIND_SET = new Set(GRAPHIC_KINDS);
const ROOT_KEYS = Object.freeze([
  'schema_version', 'opener', 'visual_policy', 'required_graphic',
  'music_allowed',
]);
const OPENER_KEYS = Object.freeze(['exact_text', 'max_start_seconds']);
const VISUAL_POLICY_KEYS = Object.freeze([
  'graphics_exact', 'broll_exact', 'opening_punch_required',
  'opening_visual_required', 'max_visual_gap_seconds',
]);
const REQUIRED_GRAPHIC_KEYS = Object.freeze(['kind', 'text', 'anchor_text']);

function deepFreeze(value) {
  if (value && typeof value === 'object' && !Object.isFrozen(value)) {
    Object.freeze(value);
    for (const child of Object.values(value)) deepFreeze(child);
  }
  return value;
}

// Schema-shape illustration only. Deliberately contains no source-video or
// user-specific wording; the editing prompt requires replacing every value
// from the current request and kept transcript before execution.
const CREATIVE_CONSTRAINTS_SCHEMA_EXAMPLE = deepFreeze({
  schema_version: CREATIVE_CONSTRAINTS_SCHEMA,
  opener: {
    exact_text: 'REPLACE WITH EXACT OPENER',
    max_start_seconds: 3,
  },
  visual_policy: {
    graphics_exact: 1,
    broll_exact: 0,
    opening_punch_required: true,
    opening_visual_required: false,
    max_visual_gap_seconds: null,
  },
  required_graphic: {
    kind: 'callout',
    text: 'REPLACE DISPLAY COPY',
    anchor_text: 'REPLACE WITH EXACT KEPT ANCHOR',
  },
  music_allowed: false,
});

function isPlainObject(value) {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return false;
  const prototype = Object.getPrototypeOf(value);
  return prototype === Object.prototype || prototype === null;
}

function exactKeys(value, expected) {
  if (!isPlainObject(value)) return false;
  const actual = Object.keys(value).sort();
  const wanted = [...expected].sort();
  return actual.length === wanted.length &&
    actual.every((key, index) => key === wanted[index]);
}

function canonicalString(value, label, maximum, requireCanonical) {
  if (typeof value !== 'string') {
    throw new TypeError(`${label} must be a string`);
  }
  const clean = canonicalText(value);
  if (!clean || clean.length > maximum || (requireCanonical && clean !== value)) {
    throw new Error(`${label} must be canonical text containing 1-${maximum} characters`);
  }
  return clean;
}

function boundedNumber(value, label, low, high) {
  if (typeof value !== 'number' || !Number.isFinite(value) ||
      value < low || value > high) {
    throw new Error(`${label} must be a finite number from ${low} to ${high}`);
  }
  return value;
}

function exactCount(value, label) {
  if (!Number.isSafeInteger(value) || value < 0 || value > 16) {
    throw new Error(`${label} must be an integer from 0 to 16`);
  }
  return value;
}

function wordTokens(value) {
  return (canonicalText(value)
    .replace(/[\u2018\u2019\u02bc\uff07]/g, "'")
    .toLowerCase().match(/[a-z0-9']+/g) || []);
}

function assertStoryLink(constraints, storyPlan) {
  if (!isPlainObject(storyPlan) || !Array.isArray(storyPlan.keep_ranges) ||
      storyPlan.keep_ranges.length < 2) {
    throw new Error('creative constraints require a valid story plan');
  }
  const firstAnchor = canonicalText(storyPlan.keep_ranges[0]?.anchor_text);
  const opener = constraints.opener.exact_text;
  if (firstAnchor !== opener && !firstAnchor.startsWith(`${opener} `)) {
    throw new Error('creative constraints opener must exactly prefix the first kept story anchor');
  }
  if (constraints.required_graphic !== null) {
    const keptTranscript = canonicalText(storyPlan.keep_ranges
      .map((range) => range?.anchor_text).join(' '));
    if (!keptTranscript.includes(constraints.required_graphic.anchor_text)) {
      throw new Error('creative constraints required graphic anchor is absent from the kept story transcript');
    }
  }
}

function validateCreativeConstraints(raw, storyPlan = null, {
  requireCanonical = false,
} = {}) {
  if (!exactKeys(raw, ROOT_KEYS)) {
    throw new Error('creative constraints do not match the executable contract');
  }
  if (raw.schema_version !== CREATIVE_CONSTRAINTS_SCHEMA) {
    throw new Error('creative constraints schema is unsupported');
  }
  if (!exactKeys(raw.opener, OPENER_KEYS)) {
    throw new Error('creative constraints opener does not match the executable contract');
  }
  const openerText = canonicalString(
    raw.opener.exact_text, 'opener.exact_text', 160, requireCanonical);
  const openerStart = boundedNumber(
    raw.opener.max_start_seconds, 'opener.max_start_seconds', 0, 10);

  if (!exactKeys(raw.visual_policy, VISUAL_POLICY_KEYS)) {
    throw new Error('creative constraints visual policy does not match the executable contract');
  }
  const policy = raw.visual_policy;
  const graphicsExact = exactCount(
    policy.graphics_exact, 'visual_policy.graphics_exact');
  const brollExact = exactCount(policy.broll_exact, 'visual_policy.broll_exact');
  if (typeof policy.opening_punch_required !== 'boolean' ||
      typeof policy.opening_visual_required !== 'boolean') {
    throw new Error('creative constraints opening requirements must be boolean');
  }
  const maxVisualGap = policy.max_visual_gap_seconds === null ? null
    : boundedNumber(policy.max_visual_gap_seconds,
      'visual_policy.max_visual_gap_seconds', 1, 300);

  let requiredGraphic = null;
  if (raw.required_graphic !== null) {
    if (!exactKeys(raw.required_graphic, REQUIRED_GRAPHIC_KEYS)) {
      throw new Error('creative constraints required graphic does not match the executable contract');
    }
    if (graphicsExact !== 1) {
      throw new Error('creative constraints required graphic requires graphics_exact=1');
    }
    const rawKind = canonicalString(
      raw.required_graphic.kind, 'required_graphic.kind', 16, requireCanonical);
    const kind = rawKind.toLowerCase();
    if (requireCanonical && kind !== rawKind) {
      throw new Error('required_graphic.kind must be canonical lowercase text');
    }
    if (!GRAPHIC_KIND_SET.has(kind)) {
      throw new Error('creative constraints required graphic kind is unsupported');
    }
    const text = canonicalString(
      raw.required_graphic.text, 'required_graphic.text', 44, requireCanonical);
    const words = text.match(/[A-Za-z0-9']+/g) || [];
    if (text !== text.toUpperCase() || words.length > 4) {
      throw new Error('required_graphic.text must be uppercase and at most four words');
    }
    const anchorText = canonicalString(
      raw.required_graphic.anchor_text, 'required_graphic.anchor_text', 200,
      requireCanonical);
    const anchorWords = wordTokens(anchorText);
    if (anchorWords.length < 5 || anchorWords.length > 20) {
      throw new Error('required_graphic.anchor_text must contain 5-20 exact words');
    }
    requiredGraphic = Object.freeze({ kind, text, anchor_text: anchorText });
  } else if (graphicsExact === 1) {
    throw new Error('creative constraints with one exact graphic require its specification');
  }

  if (typeof raw.music_allowed !== 'boolean') {
    throw new Error('creative constraints music_allowed must be boolean');
  }

  const constraints = Object.freeze({
    schema_version: CREATIVE_CONSTRAINTS_SCHEMA,
    opener: Object.freeze({
      exact_text: openerText,
      max_start_seconds: openerStart,
    }),
    visual_policy: Object.freeze({
      graphics_exact: graphicsExact,
      broll_exact: brollExact,
      opening_punch_required: policy.opening_punch_required,
      opening_visual_required: policy.opening_visual_required,
      max_visual_gap_seconds: maxVisualGap,
    }),
    required_graphic: requiredGraphic,
    music_allowed: raw.music_allowed,
  });
  if (storyPlan !== null) assertStoryLink(constraints, storyPlan);
  return constraints;
}

function normalizeCreativeConstraints(raw, storyPlan) {
  try {
    return validateCreativeConstraints(raw, storyPlan);
  } catch (_) {
    return null;
  }
}

function structuralCreativeConstraints(raw, storyPlan) {
  validateCreativeConstraints(raw, storyPlan, { requireCanonical: true });
  return raw;
}

module.exports = Object.freeze({
  CREATIVE_CONSTRAINTS_SCHEMA,
  GRAPHIC_KINDS,
  CREATIVE_CONSTRAINTS_SCHEMA_EXAMPLE,
  validateCreativeConstraints,
  normalizeCreativeConstraints,
  structuralCreativeConstraints,
});

File: autoeditor/creative_constraints.py
"""Closed, user-approved creative constraints for a premium edit.

The prose brief remains useful director context, but it is not executable.
This module is the small deterministic contract that makes visual promises
machine-checkable before planning and again before release.
"""
from __future__ import annotations

import hashlib
import json
import math
import re
import unicodedata


SCHEMA_VERSION = "autoeditor-creative-constraints/v1"
GRAPHIC_KINDS = frozenset({"keyword", "stat", "callout", "bars"})
_ROOT_KEYS = frozenset({
    "schema_version", "opener", "visual_policy", "required_graphic",
    "music_allowed",
})
_OPENER_KEYS = frozenset({"exact_text", "max_start_seconds"})
_VISUAL_KEYS = frozenset({
    "graphics_exact", "broll_exact", "opening_punch_required",
    "opening_visual_required", "max_visual_gap_seconds",
})
_GRAPHIC_KEYS = frozenset({"kind", "text", "anchor_text"})
_APOSTROPHE_TRANSLATION = str.maketrans({
    "\u2018": "'", "\u2019": "'", "\u02bc": "'", "\uff07": "'",
})


class CreativeConstraintsError(ValueError):
    """The approved creative policy is malformed or internally inconsistent."""


def canonical_text(value: object) -> str:
    return re.sub(
        r"\s+", " ", unicodedata.normalize("NFKC", str(value or ""))
    ).strip()


def word_tokens(value: object) -> list[str]:
    """Use one token boundary contract for constraints and creative QA."""
    text = canonical_text(value).translate(_APOSTROPHE_TRANSLATION).lower()
    return re.findall(r"[a-z0-9']+", text)


def _plain_dict(value: object, keys: frozenset[str], label: str) -> dict:
    if not isinstance(value, dict) or set(value) != keys:
        raise CreativeConstraintsError(
            f"{label} must contain exactly: {', '.join(sorted(keys))}"
        )
    return value


def _bounded_number(value: object, label: str, low: float,
                    high: float) -> float:
    if (isinstance(value, bool) or not isinstance(value, (int, float))
            or not math.isfinite(float(value))
            or not low <= float(value) <= high):
        raise CreativeConstraintsError(
            f"{label} must be a finite number from {low:g} to {high:g}"
        )
    return float(value)


def _exact_count(value: object, label: str) -> int:
    if (isinstance(value, bool) or not isinstance(value, int)
            or not 0 <= value <= 16):
        raise CreativeConstraintsError(f"{label} must be an integer from 0 to 16")
    return value


def validate_creative_constraints(raw: object) -> dict:
    root = _plain_dict(raw, _ROOT_KEYS, "creative constraints")
    if root["schema_version"] != SCHEMA_VERSION:
        raise CreativeConstraintsError("creative constraints schema is unsupported")

    opener = _plain_dict(root["opener"], _OPENER_KEYS, "opener")
    opener_text = canonical_text(opener["exact_text"])
    if not opener_text or len(opener_text) > 160:
        raise CreativeConstraintsError("opener.exact_text must contain 1-160 characters")
    opener_start = _bounded_number(
        opener["max_start_seconds"], "opener.max_start_seconds", 0, 10
    )

    policy = _plain_dict(
        root["visual_policy"], _VISUAL_KEYS, "visual_policy"
    )
    graphics_exact = _exact_count(
        policy["graphics_exact"], "visual_policy.graphics_exact"
    )
    broll_exact = _exact_count(
        policy["broll_exact"], "visual_policy.broll_exact"
    )
    for key in ("opening_punch_required", "opening_visual_required"):
        if not isinstance(policy[key], bool):
            raise CreativeConstraintsError(f"visual_policy.{key} must be boolean")
    raw_gap = policy["max_visual_gap_seconds"]
    max_gap = None if raw_gap is None else _bounded_number(
        raw_gap, "visual_policy.max_visual_gap_seconds", 1, 300
    )

    required = root["required_graphic"]
    clean_graphic = None
    if required is not None:
        graphic = _plain_dict(required, _GRAPHIC_KEYS, "required_graphic")
        if graphics_exact != 1:
            raise CreativeConstraintsError(
                "required_graphic requires visual_policy.graphics_exact=1"
            )
        kind = canonical_text(graphic["kind"]).lower()
        text = canonical_text(graphic["text"])
        anchor = canonical_text(graphic["anchor_text"])
        if kind not in GRAPHIC_KINDS:
            raise CreativeConstraintsError("required_graphic.kind is unsupported")
        if (not text or len(text) > 44 or text != text.upper()
                or len(re.findall(r"[A-Za-z0-9']+", text)) > 4):
            raise CreativeConstraintsError(
                "required_graphic.text must be uppercase, at most 44 characters "
                "and four words"
            )
        if (not anchor or len(anchor) > 200
                or not 5 <= len(word_tokens(anchor)) <= 20):
            raise CreativeConstraintsError(
                "required_graphic.anchor_text must contain 5-20 exact words "
                "and at most 200 characters"
            )
        clean_graphic = {"kind": kind, "text": text, "anchor_text": anchor}
    elif graphics_exact == 1:
        raise CreativeConstraintsError(
            "one exact graphic requires a required_graphic specification"
        )

    if not isinstance(root["music_allowed"], bool):
        raise CreativeConstraintsError("music_allowed must be boolean")

    return {
        "schema_version": SCHEMA_VERSION,
        "opener": {
            "exact_text": opener_text,
            "max_start_seconds": opener_start,
        },
        "visual_policy": {
            "graphics_exact": graphics_exact,
            "broll_exact": broll_exact,
            "opening_punch_required": policy["opening_punch_required"],
            "opening_visual_required": policy["opening_visual_required"],
            "max_visual_gap_seconds": max_gap,
        },
        "required_graphic": clean_graphic,
        "music_allowed": root["music_allowed"],
    }


def canonical_json(value: object) -> str:
    return json.dumps(
        validate_creative_constraints(value), ensure_ascii=True,
        sort_keys=True, separators=(",", ":"), allow_nan=False,
    )


def constraints_sha256(value: object) -> str:
    return hashlib.sha256(canonical_json(value).encode("utf-8")).hexdigest()
